partner_affiliate/models/res_partner.py

```diff
@@ -23,3 +23,31 @@ class ResPartner(models.Model):
                                     string='Affiliates',
                                     domain=[('active', '=', True),
                                             ('is_company', '=', True)])
+
+    @api.onchange('parent_id')
+    def onchange_parent_id(self):
+        # Keep the original address info to set it back if is a company.
+        original_address = self.get_original_address()
+
+        new_partner = super(ResPartner, self).onchange_parent_id()
+
+        # When the affiliate is a company, we must set back its address
+        # because the super call changes its address by the new parent address.
+        # In addition, the type must be set to affiliate instead of contact.
+        if self.is_company:
+            self.type = 'affiliate'
+            self.update_address(original_address)
+            return
+
+        return new_partner
+
+    def get_original_address(self):
```

Comment:
This function should be defined before the onchange, as it is used there

Reply:
Ok.